### PR TITLE
add snapshot,expand api

### PR DESCRIPTION
**Pull request by peng9808**
#### What this PR does / why we need it:
https://github.com/hwameistor/hwameistor/issues/1210
#### Special notes for your reviewer:
NONE
#### Does this PR introduce a user-facing change?
add Snapshot display and volume expansion

```release-note

```

### CODE REVIEW

File: pkg/apiserver/controller/SnapshotController.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+	var queryPage hwameistorapi.QueryPage
+	queryPage.Page = int32(p)
+	queryPage.PageSize = int32(ps)
+	queryPage.SnapshotName = snapshotName
+	queryPage.SnapshotState = hwameistorapi.VolumeStatefuzzyConvert(state)
+	queryPage.VolumeName = volumeName
+
+	lvs, err := v.m.SnapshotController().ListLocalSnapshot(queryPage)
+	if err != nil {
+		failRsp.ErrCode = 500
```
Comment (SSmallMonster) [inline pkg/apiserver/controller/SnapshotController.go:68]: there is defination about http status at package `http`, use that may be better

File: pkg/apiserver/manager/hwameistor/localsnapshot_controller.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+			((queryPage.SnapshotName == "") && (queryPage.SnapshotState == apisv1alpha1.VolumeStateEmpty) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) ||
+			((queryPage.SnapshotName != "" && strings.Contains(vol.Name, queryPage.SnapshotName)) && (queryPage.SnapshotState != "" && queryPage.SnapshotState == vol.Status.State) && (queryPage.VolumeName == "")) ||
+			((queryPage.SnapshotName != "" && strings.Contains(vol.Name, queryPage.SnapshotName)) && (queryPage.SnapshotState == apisv1alpha1.VolumeStateEmpty) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) ||
+			((queryPage.SnapshotName == "") && (queryPage.SnapshotState != "" && queryPage.SnapshotState == vol.Status.State) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) ||
+			((queryPage.SnapshotName != "" && strings.Contains(vol.Name, queryPage.SnapshotName)) && (queryPage.SnapshotState != "" && queryPage.SnapshotState == vol.Status.State) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) {
```
Comment (SSmallMonster) [inline pkg/apiserver/manager/hwameistor/localsnapshot_controller.go:75]: logic here is too complex, may be we can try to make it easy to read

File: pkg/apiserver/manager/hwameistor/metric_controller.go [hunk 3]
```diff
@@ -520,3 +520,64 @@ func (mController *MetricController) convertNodeStorageUseMetric(storagepoolclas
...
+		//	((queryPage.Name == "") && (queryPage.SnapshotState == apisv1alpha1.VolumeStateEmpty) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) ||
+		//	((queryPage.Name != "" && strings.Contains(vol.Name, queryPage.SnapshotName)) && (queryPage.SnapshotState != "" && queryPage.SnapshotState == vol.Status.State) && (queryPage.VolumeName == "")) ||
+		//	((queryPage.Name != "" && strings.Contains(vol.Name, queryPage.SnapshotName)) && (queryPage.SnapshotState == apisv1alpha1.VolumeStateEmpty) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) ||
+		//	((queryPage.Name == "") && (queryPage.SnapshotState != "" && queryPage.SnapshotState == vol.Status.State) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) ||
+		//	((queryPage.Name != "" && strings.Contains(vol.Name, queryPage.SnapshotName)) && (queryPage.SnapshotState != "" && queryPage.SnapshotState == vol.Status.State) && (queryPage.VolumeName != "" && queryPage.VolumeName == vol.Spec.SourceVolume)) {
+		//
+		//	snaps = append(snaps, vol)
+		//}
```
Comment (SSmallMonster) [inline pkg/apiserver/manager/hwameistor/metric_controller.go:575]: useless codes can be moved out

File: pkg/apiserver/router/router.go [hunk 4]
```diff
@@ -58,12 +59,23 @@ func CollectRoute(r *gin.Engine) *gin.Engine {
 	v1.GET("/cluster/volumes/:volumeName/convert", volumeController.GetVolumeConvertOperation)
 	v1.POST("/cluster/volumes/:volumeName/convert", volumeController.VolumeConvertOperation)
 
+	//volumes expand
+	v1.GET("/cluster/volumes/:volumeName/expand", volumeController.GetVolumeExpandOperation)
+	v1.POST("/cluster/volumes/:volumeName/expand", volumeController.VolumeExpandOperation)
+
 	v1.GET("/cluster/volumes/:volumeName/operations", volumeController.VolumeOperationGet)
+	//volumes snapshot
+	v1.GET("/cluster/volumes/:volumeName/snapshot", volumeController.VolumeSnapshotList)
```
Comment (SSmallMonster) [inline pkg/apiserver/router/router.go:68]: using plural `snapshots`

**Review by SSmallMonster — COMMENTED** [review]
overall lgtm

File: pkg/apiserver/controller/SnapshotController.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+// @Param       state query string false "state"
+// @Param       volumeName query string false "volumeName"
+// @Accept      json
+// @Produce     json
+// @Success     200 {object}  api.SnapshotList
+// @Router      /cluster/snapshots [get]
+func (v *SnapshotController) SnapshotList(ctx *gin.Context) {
+	var failRsp hwameistorapi.RspFailBody
+
+	// 获取path中的page
```
[the remark this answers is not in the document]
Comment (peng9808) [reply pkg/apiserver/controller/SnapshotController.go]: It has been changed to English

File: pkg/apiserver/controller/SnapshotController.go (hunk 1 above)
Comment (peng9808) [reply pkg/apiserver/controller/SnapshotController.go]: Has been modified

File: pkg/apiserver/manager/hwameistor/metric_controller.go (hunk 3 above)
Comment (peng9808) [reply pkg/apiserver/manager/hwameistor/metric_controller.go]: Has been  moved out

**Review by SSmallMonster — APPROVED** [review]

**Pull request merged**